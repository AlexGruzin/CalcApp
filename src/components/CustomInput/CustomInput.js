import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './CustomInput.styles.less';

const CustomInput = React.forwardRef((props, ref) => {
  const { type, name, className, label, isError, errMessage, beforeText, ...rest } = props;

  return (
    <div className={classNames(className, 'custom-input', { withLabel: label, isError })}>
      {label && <label htmlFor={name}>{label}</label>}
      {beforeText && <span className="custom-input__before">{beforeText}</span>}
      <input {...rest} aria-label={name} name={name} id={name} ref={ref} type={type} />
      {/* // TODO make error appear animation */}
      <div className="custom-input__error">{isError && errMessage ? errMessage : null}</div>
    </div>
  );
});

CustomInput.propTypes = {
  type: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.array, PropTypes.func]),
  label: PropTypes.string,
  name: PropTypes.string,
  className: PropTypes.string,
  isError: PropTypes.bool,
  beforeText: PropTypes.string,
  errMessage: PropTypes.string,
};

CustomInput.defaultProps = {
  isError: false,
  children: null,
  type: 'text',
  label: '',
  name: '',
  className: '',
  beforeText: '',
  errMessage: '',
};

export default CustomInput;
