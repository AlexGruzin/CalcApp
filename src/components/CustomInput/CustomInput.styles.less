@import "~styles/generic.less";

.custom-input {
  position: relative;
}

.custom-input input {
  display: flex;
  flex-direction: column;
  padding: 3px 12px;

  .montserrat-medium();
  color: var(--grey700);
  font-size: var(--fontSize-tiny);
}

.custom-input.isError input {
  border-color: var(--error-border);
}

.custom-input.withLabel input {
  margin-top: var(--space-x3);
}

.custom-input input:focus::placeholder {
  color: transparent;
}

.custom-input input:focus,input:active {
  box-shadow: none;
  outline: 0;
}

.custom-input:not(.isError) input:focus,.custom-input:not(.isError) input:active {
  border: 1px solid var(--outline-default);
}

.custom-input input::placeholder {
  text-align: center;
  padding: 3px 10px;
  text-transform: capitalize;
  .montserrat-medium();
  color: var(--grey700);
  opacity: 1;
  font-size: var(--fontSize-tiny);
}

.custom-input__before {
  display: flex;
  align-items: center;
  position: absolute;
  top: 0;
  padding: 3px 6px;
  border: 1px solid transparent;
  line-height: normal;
  height: 28px;
  bottom: 0;
  .montserrat-medium();
  color: var(--grey700);
}

.custom-input__error {
  height: 11px;
  min-height: 11px;
  margin: var(--space-x1) var(--space-x2) 0 var(--space-x2);
  .montserrat-medium();
  color: var(--error-color);
  font-size: var(--fontSize-superbTiny);
  letter-spacing: 0.18px;
}
