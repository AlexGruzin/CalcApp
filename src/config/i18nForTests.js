import i18n from 'i18next';

// locales
import US from '../locales/en';

i18n.init({
  debug: false,
  fallbackLng: ['US'],
  lng: 'US',

  ns: ['auth', 'resetPassword'],
  defaultNS: 'auth',

  resources: {
    US,
  },
});

export default i18n;

