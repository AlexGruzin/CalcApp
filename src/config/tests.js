import React from 'react';
import { render } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next'
import i18n from "./i18nForTests";
import jsdom from 'jsdom';
const { JSDOM } = jsdom;

/*
 * jest-dom
 * https://github.com/testing-library/jest-dom
 */
import '@testing-library/jest-dom/extend-expect';

/*
 * jest-fetch-mock
 * https://github.com/jefflau/jest-fetch-mock
 */
require('jest-fetch-mock').enableMocks();

if (global.document) {
  global.document.createRange = () => ({
    setStart: () => {
    },
    setEnd: () => {
    },
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
    createContextualFragment: str => JSDOM.fragment(str),
    // createContextualFragment: str => document.createElement('hr'),
  });
  global.open = () => undefined;
}

export const renderWithi18next = Component => {
  const Comp = React.cloneElement(Component, {
    changeLanguage: lng => {
      i18n.changeLanguage(lng);
      rerender(<I18nextProvider i18n={i18n}>{Comp}</I18nextProvider>);
    }
  });
  const defaultRender = render(
    <I18nextProvider i18n={i18n}>{Component}</I18nextProvider>
  );
  const { rerender } = defaultRender;
  return defaultRender;
};


// /**
//  * Wraps a component with an Intl Provider.
//  * @param {Object} component The component to render
//  */
// const withTestIntl = component => <IntlProvider locale="en">{component}</IntlProvider>;
//
// /*
//  * https://testing-library.com/docs/example-react-intl
//  */
// export const renderWithReactIntl = component => render(withTestIntl(component));

