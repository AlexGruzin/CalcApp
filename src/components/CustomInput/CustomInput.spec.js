import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { renderWithi18next } from 'config/tests';

import CustomInput from './CustomInput';

describe('CustomInput tests', () => {
  it('Check snapshots of CustomInput with value are equal.', () => {
    const { asFragment } = render(<CustomInput value="test" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Input placeholder rendered', () => {
    const { getByPlaceholderText } = render(<CustomInput placeholder="placeHolded" />);

    const input = getByPlaceholderText('placeHolded');
    expect(input).toBeInTheDocument();
  });

  it('CustomInput label rendered', () => {
    const { getByText } = renderWithi18next(<CustomInput label="customLabel" />);

    const input = getByText('customLabel');
    expect(input).toBeInTheDocument();
  });

  it('New value type working', async () => {
    const mockOnChange = jest.fn();
    const { getByLabelText } = renderWithi18next(<CustomInput onChange={mockOnChange} name="name" label="custom" />);

    const input = getByLabelText('custom');

    await userEvent.type(input, '123');

    expect(mockOnChange).toHaveBeenCalledTimes(3);
  });
});
