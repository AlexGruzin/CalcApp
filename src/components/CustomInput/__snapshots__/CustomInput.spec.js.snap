// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomInput tests Check snapshots of CustomInput with value are equal. 1`] = `
<DocumentFragment>
  <div
    class="custom-input"
  >
    <input
      aria-label=""
      id=""
      name=""
      type="text"
      value="test"
    />
    <div
      class="custom-input__error"
    />
  </div>
</DocumentFragment>
`;
